### PR TITLE
Gitlab service

### PR DESCRIPTION
**Pull request by tmcgilchrist**
This PR sets up a second pipeline for projects hosted on gitlab.com and adds support for it to the Dream based UI.

TODO:
 * [x] support for Badges on gitlab hosted projects
 * [ ] require ocurrent support for setting --current-state-dir https://github.com/ocurrent/ocurrent/pull/326 OR
 * [ ] create a docker-compose.yml file for testing both pipelines 
 * [x] Deployer for new pipeline https://github.com/ocurrent/ocurrent-deployer/pull/134

### CODE REVIEW

File: gitlab/main.ml [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    Lwt.return (vat, Some rpc_engine_resolver)
+
+module Gitlab = struct
+  (* Access control policy. *)
+  let has_role user role =
+    match user with
+    | None -> role = `Viewer              (* Unauthenticated users can only look at things. *)
+    | Some user ->
+      match Current_web.User.id user, role with
+      | "gitlab:tmcgilchrist", _ -> true  (* This user has all roles *)
```
Comment (novemberkilo) [inline gitlab/main.ml:95]: Figuring this is interim? :)

File: web-ui/controller/git_forge.ml [hunk 2]
```diff
@@ -0,0 +1,202 @@
+module type View = View.Git_forge.View
+
+module Client = Ocaml_ci_api.Client
+module type Controller = sig
```
Comment (novemberkilo) [inline web-ui/controller/git_forge.ml:4]: 👍🏾

File: web-ui/view/git_forge.ml [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+(* Common utility functions for Git_forge views. *)
+let short_hash = Astring.String.with_range ~len:6
+
+let rec intersperse ~sep = function
+  | [] -> []
+  | [ x ] -> [ x ]
+  | x :: xs -> x :: sep :: intersperse ~sep xs
+
+
+(* Common view partials for Git_forge views. *)
```
Comment (novemberkilo) [inline web-ui/view/git_forge.ml:75]: 👍🏾

**Review by novemberkilo — APPROVED** [review]
Thanks @tmcgilchrist - your Module fu makes this nice and clean.

File: gitlab/main.ml (hunk 1 above)
Comment (tmcgilchrist) [reply gitlab/main.ml:95]: It is equivalent to this module for GitHub https://github.com/ocurrent/ocaml-ci/blob/master/service/github.ml#L6-L11
The main use is for the permissions checks when triggering certain actions in the web UI. That all might change with the Dream redesign.

**Pull request merged**